### PR TITLE
STORM-1778: scheme extension framework for KafkaSourceProvider and csv s…

### PR DESCRIPTION
**Pull request by chuanlei**
feature: scheme extension framework for KafkaSourceProvider and csv scheme support

### CODE REVIEW

File: external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/CsvScheme.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  public CsvScheme(List<FieldInfo> fieldsInfo) {
+    this.fieldsInfo = fieldsInfo;
+    fields = new ArrayList<>();
+    for (FieldInfo info : fieldsInfo) {
+      fields.add(info.name());
+    }
+  }
+
+  @Override
+  public List<Object> deserialize(ByteBuffer ser) {
```
Comment (HeartSaVioR) [inline external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/CsvScheme.java]: 1. I think we can have some test cases on this.
2. StringScheme#deserializeString doesn't use ByteBuffer#toString, and javadoc says it provides information about instance state, not content: https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#toString()

File: external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/KafkaDataSourcesProvider.java [hunk 2]
```diff
@@ -171,12 +171,23 @@ public ISqlTridentDataSource constructTrident(URI uri, String inputFormatClass,
       }
     }
     Preconditions.checkState(primaryIndex != -1, "Kafka stream table must have a primary key");
-    conf.scheme = new SchemeAsMultiScheme(new JsonScheme(fieldNames));
     ObjectMapper mapper = new ObjectMapper();
     Properties producerProp = new Properties();
     try {
       @SuppressWarnings("unchecked")
       HashMap<String, Object> map = mapper.readValue(properties, HashMap.class);
+      if(!map.containsKey("scheme")){
```
Comment (HeartSaVioR) [inline external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/KafkaDataSourcesProvider.java]: Please fix spaces between L179 and L187.

**Comment by HeartSaVioR** [comment]
Btw, how about adding TSV? It's similar approach to CSV so easy to address.

File: external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/CsvScheme.java [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    String str = ser.toString();
+    String[] parts = str.split(",");
```
Comment (vesense) [inline external/sql/storm-sql-kafka/src/jvm/org/apache/storm/sql/kafka/CsvScheme.java]: This is not standard CSV format, something like `a,"b,c",d` is also valid. We can use TSV format to address.

**Comment by chuanlei** [comment]
@vesense @HeartSaVioR 
Thank you for the suggestion, I will fix it asap.

**Comment by vesense** [comment]
@chuanlei I think this issue has been fixed by STORM-2173. Can you close this?
BTW, the` CsvScheme` in STORM-2173 uses standard RFC4180 CSV parser supporting strings like `a,"b,c",d`.

**Pull request closed without merging**